src/linear_algebra/lagrange.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Kenny Lau.
+-/
+
+import data.polynomial algebra.big_operators
+
+/-!
+# Lagrange interpolation
+
+## Main definitions
+
+* `lagrange.basis s x` where `s : finset F` and `x : F`: the Lagrange basis polynomial
+  that evaluates to `1` at `x` and `0` at other elements of `s`.
+* `lagrange.interpolate s f` where `s : finset F` and `f : s → F`: the Lagrange interpolant
+  that evaluates to `f x` at `x` for `x ∈ s`.
+
+-/
+
+noncomputable theory
+open_locale classical
+
+universe u
+
+namespace lagrange
+
+variables {F : Type u} [decidable_eq F] [field F] (s : finset F)
+
+open polynomial
+
+/-- Lagrange basis polynomials that evaluate to $1$ at `x` and $0$ at other elements of `s`. -/
+def basis (x : F) : polynomial F :=
+(s.erase x).prod $ λ y, C (x - y)⁻¹ * (X - C y)
+
+theorem eval_basis_self (x : F) : (basis s x).eval x = 1 :=
+by { rw [basis, ← (s.erase x).prod_hom (eval x), finset.prod_eq_one],
+  intros y hy, simp_rw [eval_mul, eval_sub, eval_C, eval_X],
+  exact inv_mul_cancel (sub_ne_zero_of_ne (finset.ne_of_mem_erase hy).symm) }
+
+theorem eval_basis_ne (x y : F) (h1 : y ∈ s) (h2 : y ≠ x) : (basis s x).eval y = 0 :=
+by { rw [basis, ← (s.erase x).prod_hom (eval y), finset.prod_eq_zero (finset.mem_erase.2 ⟨h2, h1⟩)],
+  simp_rw [eval_mul, eval_sub, eval_C, eval_X, sub_self, mul_zero] }
+
+theorem eval_basis (x y : F) (h : y ∈ s) : (basis s x).eval y = if y = x then 1 else 0 :=
+by { split_ifs with H, { subst H, apply eval_basis_self }, { exact eval_basis_ne s x y h H } }
+
+theorem nat_degree_basis (x : F) (hx : x ∈ s) : (basis s x).nat_degree = s.card - 1 :=
+by { unfold basis, generalize hsx : s.erase x = sx,
```

Comment:
Please use `begin .. end` for multiline proofs.

Reply:
done